### ISSUE TITLE
Regression, theming inconsistency in 190c

### ISSUE THREAD

**Issue by hollunder** [pkulchenko/ZeroBraneStudio#1024]
Hi Paul,
I just noticed something weird. Have a look at these screenshots.

![2019-11-24-004528_3840x1080_scrot](https://user-images.githubusercontent.com/8207622/69487167-3dc7db00-0e4d-11ea-9c5c-6c7e2430e77e.png)
![2019-11-24-005445_3840x1080_scrot](https://user-images.githubusercontent.com/8207622/69487165-3dc7db00-0e4d-11ea-92ad-0191e49bda47.png)

The second one where it's just 1.80 is how I'd expect it to look. This one was built using this script: [zerobrane-studio](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=zerobrane-studio).
They all use the same wxgtk/wxlua (my build from git).
What's really weird is that it seems to me that only the inner parts of zbstudio are so ugly while the rest looks like it is adhering to the gtk2 theme I have set.

**Comment by pkulchenko** [comment]
@hollunder, what inconsistency are we looking at here? The tabs in the notebook? There was a change in wxwidgets to reset the default configuration, so I had to enforce a specific style for the notebook, but from the documentation it should match what was there earlier.

Can you try commenting out the line 742 (`nb:SetArtProvider(wxaui.wxAuiGenericTabArt())`) in `src/editor/gui.lua` and check if it matches the "old" look? If you are referring to something else, please be specific.

**Comment by hollunder** [comment]
Yes, I was referring to the tabs and their background. I didn't notice it earlier because on my laptop I use a bright theme and there it looks fine. It's apparently only with a dark theme (Arc-Dark-solid in this case) where it falls apart (extreme gloss, barely readable font).

For me that file has only 739 lines and does not contain said line.
However, I took a pot shot and commented line 735 `nb:SetArtProvider(ide:GetTabArt())` and that got rid of the problem.
![2019-11-24-084609_3840x1080_scrot](https://user-images.githubusercontent.com/8207622/69491546-a1262d00-0e8e-11ea-91c3-5acd0580ff92.png)

Btw. because GTK3 is mentioned in the source comment there, does it work with GTK3 and do you recommend 3 over 2?

**Comment by pkulchenko** [comment]
Yes, I tested with GTK3 and didn't see any difference; I didn't realize that this is going to be different with GTK2. I think I may just have disable this for Linux to keep it consistent with how it's set as the default.

**Comment by pkulchenko** [comment]
Actually, can you try with GTK3 to see if there is any difference with the result?

**Comment by hollunder** [comment]
I just tried with GTK3, by setting wx-config-gtk3, which seems to have worked: `-- * - wxWidgets_LIBRARIES         = -pthread;;;-lwx_gtk3u_stc-3.0;-lwx_gtk3u_gl-3.0;-lwx_gtk3u_html-3.0;-lwx_gtk3u_aui-3.0;-lwx_gtk3u_adv-3.0;-lwx_gtk3u_core-3.0;-lwx_baseu_net-3.0;-lwx_baseu-3.0`

This is on my laptop where I don't have the same dark theme installed, so I picked a different dark theme, so it doesn't look identical but the issue is still clearly there.

![2019-11-26-082326_1366x768_scrot](https://user-images.githubusercontent.com/8207622/69607775-458ea780-1026-11ea-92e3-9bc0374b9fcf.png)

**Comment by pkulchenko** [comment]
This is something that needs to be addressed in wxwidgets; I opened a [ticket for it](https://trac.wxwidgets.org/ticket/18601).

**Comment by pkulchenko** [comment]
@hollunder, could you report back the numbers you get for that back theme when you execute the following commands in the Console?

```
wx.wxSystemSettings.GetColour(wx.wxSYS_COLOUR_3DFACE):GetAsString()
wx.wxSystemSettings.GetColour(wx.wxSYS_COLOUR_CAPTIONTEXT):GetAsString()
wx.wxSystemSettings.GetColour(wx.wxSYS_COLOUR_INACTIVECAPTIONTEXT):GetAsString()
```

**Comment by hollunder** [comment]
For the Breeze-Dark theme, a grey-ish one.
```
wx.wxSystemSettings.GetColour(wx.wxSYS_COLOUR_3DFACE):GetAsString()
"rgb(49, 54, 59)"
wx.wxSystemSettings.GetColour(wx.wxSYS_COLOUR_CAPTIONTEXT):GetAsString()
"rgb(239, 240, 241)"
wx.wxSystemSettings.GetColour(wx.wxSYS_COLOUR_INACTIVECAPTIONTEXT):GetAsString()
"rgb(127, 140, 141)"
```

For deepin-dark, a almost black theme.
```
wx.wxSystemSettings.GetColour(wx.wxSYS_COLOUR_3DFACE):GetAsString()
"rgb(27, 30, 32)"
wx.wxSystemSettings.GetColour(wx.wxSYS_COLOUR_CAPTIONTEXT):GetAsString()
"rgb(137, 145, 154)"
wx.wxSystemSettings.GetColour(wx.wxSYS_COLOUR_INACTIVECAPTIONTEXT):GetAsString()
"rgb(137, 145, 154)"
```

These are two I have installed on the laptop, I can send you the colours for the one I'm actually using on the desktop in a couple hours.

**Comment by hollunder** [comment]
For the theme I use on my desktop, Arc-Dark-Solid, these are the colours.
```
wx.wxSystemSettings.GetColour(wx.wxSYS_COLOUR_3DFACE):GetAsString()
"rgb(56, 60, 74)"
wx.wxSystemSettings.GetColour(wx.wxSYS_COLOUR_CAPTIONTEXT):GetAsString()
"rgb(211, 218, 227)"
wx.wxSystemSettings.GetColour(wx.wxSYS_COLOUR_INACTIVECAPTIONTEXT):GetAsString()
"rgb(211, 218, 227)"
```

**Comment by pkulchenko** [comment]
@hollunder, thank you for the details. Can you apply the following patch to wxwidgets and recompile it? I'm not sure if it's going to apply it cleanly to your version of wxwidgets (3.0.4?), but it should explicitly set the font color in DrawTab method of `wxAuiGenericTabArt`. You can ignore the first fragment if it doesn't patch cleanly.

```
diff --git a/src/aui/tabart.cpp b/src/aui/tabart.cpp
index 7c512f95d7..c659534568 100644
--- a/src/aui/tabart.cpp
+++ b/src/aui/tabart.cpp
@@ -267,14 +267,6 @@ void wxAuiGenericTabArt::DrawBackground(wxDC& dc,
     // draw background
     int topLightness = 90;
     int bottomLightness = 170;
-    if ((m_baseColour.Red() < 75)
-        && (m_baseColour.Green() < 75)
-        && (m_baseColour.Blue() < 75))
-    {
-        //dark mode, we cannot go very light
-        topLightness = 90;
-        bottomLightness = 110;
-    }
 
     wxColor top_color    = m_baseColour.ChangeLightness(topLightness);
     wxColor bottom_color = m_baseColour.ChangeLightness(bottomLightness);
@@ -585,16 +577,9 @@ void wxAuiGenericTabArt::DrawTab(wxDC& dc,
                           tab_width - (text_offset-tab_x) - close_button_width);
 
     // draw tab text
-#if defined( __WXMAC__ )
-    if (page.active)
-    {
-        dc.SetTextForeground(wxSystemSettings::GetColour(wxSYS_COLOUR_CAPTIONTEXT));
-    }
-    else
-    {
-        dc.SetTextForeground(wxSystemSettings::GetColour(wxSYS_COLOUR_INACTIVECAPTIONTEXT));
-    }
-#endif
+    wxColor font_color = wxSystemSettings::GetColour(
+        page.active ? wxSYS_COLOUR_CAPTIONTEXT : wxSYS_COLOUR_INACTIVECAPTIONTEXT);
+    dc.SetTextForeground(font_color);
     dc.DrawText(draw_text,
                 text_offset,
                 drawn_tab_yoff + (drawn_tab_height)/2 - (texty/2) - 1);
```

Please let me know the results with one of the dark themes.

**Comment by hollunder** [comment]
I figure when I have to build wxgtk myself I might as well use a newer version and built 3.1.3. The patch applied cleanly, but wxLua doesn't build against that.
With the gtk2 version it fails pretty much immediately because it can't find some headers, with the gtk2 version during linking.
I'll require a little more time to sort that out. I guess I'll try and patch and build the stable version.

**Comment by hollunder** [comment]
Unfortunately the patch fails quite a bit against 3.0.4.
```
patching file src/common/appbase.cpp
Hunk #1 succeeded at 766 with fuzz 2 (offset 342 lines).
patching file src/aui/tabart.cpp
Hunk #1 FAILED at 267.
Hunk #2 FAILED at 585.
2 out of 2 hunks FAILED -- saving rejects to file src/aui/tabart.cpp.rej
```
I finally came around to modifying the patch for 3.0.4.. Sorry, I'm not used to working this way and I probably do this in overly complicate ways, hence it took me a bit.
```
--- a/src/aui/tabart.cpp	2018-03-07 17:55:38.000000000 +0100
+++ b/src/aui/tabart.cpp	2019-12-03 17:07:13.487259312 +0100
@@ -532,6 +532,9 @@
                           tab_width - (text_offset-tab_x) - close_button_width);
 
     // draw tab text
+    wxColor font_color = wxSystemSettings::GetColour(
+	page.active ? wxSYS_COLOUR_CAPTIONTEXT : wxSYS_COLOUR_INACTIVECAPTIONTEXT);
+	dc.SetTextForeground(font_color);
     dc.DrawText(draw_text,
                 text_offset,
                 drawn_tab_yoff + (drawn_tab_height)/2 - (texty/2) - 1);
```
Attached is a screenshot of that version with the theme Breeze-Dark.
![2019-12-03-185106_1366x768_scrot](https://user-images.githubusercontent.com/8207622/70075606-6eea9d00-15fd-11ea-9683-2dc1cde5e536.png)

**Comment by pkulchenko** [comment]
No problem; the screenshot gives me the information I need, thank you!

**Comment by pkulchenko** [comment]
@hollunder, the colors should be fixed in the latest release beta, but the wxwidgets changes are not likely to be ported back to 3.0.4 ;(. Can you try this branch (https://github.com/pkulchenko/ZeroBraneStudio/tree/upgrade-binaries-190d) and let me know if the tab text colors look correct? It should also detect the "dark" theme, so if you are setting the color scheme manually in the config, you can remove that (although you may still need it if your dark scheme is different from the default one.

**Comment by hollunder** [comment]
Arc-Dark-Solid
![2020-01-11_114820](https://user-images.githubusercontent.com/8207622/72203146-0692a500-3460-11ea-945e-a7b8ad603499.png)

Adwaita-Dark (Adwaita is a relatively common theme, a Gnome default I believe)
![2020-01-11_115143](https://user-images.githubusercontent.com/8207622/72203165-612c0100-3460-11ea-97db-92c9e12b03e8.png)

BlackMATE
![2020-01-11_115340](https://user-images.githubusercontent.com/8207622/72203185-ad774100-3460-11ea-9518-79ea5c46908f.png)

Breeze-Dark
![2020-01-11_115434](https://user-images.githubusercontent.com/8207622/72203203-cb44a600-3460-11ea-94eb-37dc6059ab0a.png)

HighContrastInverse (most likely intended as accessibility option)
![2020-01-11_115543](https://user-images.githubusercontent.com/8207622/72203239-efa08280-3460-11ea-9d1d-a62898100d1f.png)

Vertex-Dark
![2020-01-11_115710](https://user-images.githubusercontent.com/8207622/72203256-25456b80-3461-11ea-83a0-040cf7ad6591.png)

Menta (just to check that light ones still work)
![2020-01-11_115749](https://user-images.githubusercontent.com/8207622/72203269-427a3a00-3461-11ea-9dd1-f52313c9c3ed.png)

I hope that selection covers dark themes sufficiently. Looks good on my side.
Thanks for the fix Paul!

**Comment by hollunder** [comment]
Oh, I forgot to remove the `styles` color scheme setting before making the screenshots. I tested that briefly with the themes above and it seems to work fine.

**Comment by hollunder** [comment]
As a sidenote, a think I noticed is that the tab background now always has a gradient that it didn't have in 1.80. It sticks out quite a bit with dark themes. I suppose it's there to provide some contrast, but IMHO it's too strong.

**Comment by pkulchenko** [comment]
@hollunder, thank you for checking with all these examples! Looks good.

I agree on the gradient, but I think it was tweaked in wxwidgets to increase contrast, as you said. It also depends on the background color, so may be too strong when the color is close to 0,0,0.

**Comment by hollunder** [comment]
Oh, another thing I noticed just now. There is a very weird halo-like effect on the right side of the editor. It is very noticeable on HighContrastInverse but also visible with the other dark schemes. Hard to imagine that it's intentional. I wonder how I didn't notice this earlier...

**Comment by pkulchenko** [comment]
Do you have a screenshot? I haven't seen this one and can't image what it would be outside of the main frame.

**Comment by hollunder** [comment]
Oh I mean this screenshot as the obvious example. It does not extend outside of the program, it's just this weird effect left of the editor window scrollbar. It doesn't seem to appear next to other scroll bars (project, outline, local console).
 
HighContrastInverse (most likely intended as accessibility option)
![2020-01-11_115543](https://user-images.githubusercontent.com/8207622/72203239-efa08280-3460-11ea-9d1d-a62898100d1f.png)

**Comment by pkulchenko** [comment]
Very interesting; not sure what to say or where it's coming from. This is the first time I see this. It does seem to be limited to just that window. It could be some sort of accessibility option, but can't confirm, as I'm not sure what could be triggering it. Do "older" versions of the IDE (like the master branch) show it with the same color scheme?

**Comment by hollunder** [comment]
![2020-01-12-104553_3840x1080_scrot](https://user-images.githubusercontent.com/8207622/72217002-9009ac00-3520-11ea-9d85-7c8459727278.png)
This is 1.80 with the same scheme.
The second image in the first post is also 1.80 with a dark scheme, so there it does not happen.
Pretty much the entire series of dark screenshots above shows this effect, but usually it is more subtle. Only on the BlackMate screenshot it seems to be absent. There the scrollbar itself has a gradient, but that is part of the theme and also appears in other applications.

It gets weirder. At some point when scrolling down the vertical scroll bar appears and that effect disappears.
![2020-01-12_110616](https://user-images.githubusercontent.com/8207622/72217201-32c32a00-3523-11ea-8b83-3731c866f8ee.png)
![2020-01-12_110252](https://user-images.githubusercontent.com/8207622/72217162-bfb9b380-3522-11ea-9231-a212e0905628.png)
![2020-01-12_110303](https://user-images.githubusercontent.com/8207622/72217165-c34d3a80-3522-11ea-8034-f712c738d329.png)

I don't quite know what to make of this.

**Comment by hollunder** [comment]
Another thing I noticed, and this one is I suppose more of a related feature request, is that the tooltip remains white even if a dark theme is detected. It would make sense to have a bright font on dark background for these as well.
![2020-01-12-163822_3840x1080_scrot](https://user-images.githubusercontent.com/8207622/72221360-daefe780-3551-11ea-86cf-865e9a6a42f2.png)

**Comment by pkulchenko** [comment]
Agree on the tooltip color. This should already be fixed in `upgrade-binaries-190d` branch.

**Comment by hollunder** [comment]
No, this screenshot is from upgrade-binaries-190d.
Btw. can you explain your naming scheme there? What is the difference between all-binaries and upgrade-binaries?

**Comment by pkulchenko** [comment]
`all-binaries-190x` includes `upgrade-binaries-190x` plus luasec binaries plus changes in the master branch. It's basically a release candidate, but I'll have `all-binaries-190e` with all the recent changes shortly.

**Comment by pkulchenko** [comment]
Strange about the tooltip. I do see it colored according to the current color scheme. I'll check on Linux.

**Comment by pkulchenko** [comment]
@hollunder, can  you open `cfg/scheme-picker.lua` and try selecting some colors. The tooltip should change according to the updated background. Does it not happen for you?

How do you select the dark theme in the IDE?

**Comment by hollunder** [comment]
Usually I just look up the name in `cfg/tomorrow.lua` and modify the existing lines `user.lua`.
To test this now on my laptop I switched from my usual bright to a dark GTK scheme.
This first image is with the automatically set dark test color.
![2020-01-13-074939_1366x768_scrot](https://user-images.githubusercontent.com/8207622/72237439-c0ae1c00-35da-11ea-9ed2-e2527372fb47.png)
Then I turned to `cfg/scheme-picker.lua`, still no change.
![2020-01-13-075448_1366x768_scrot](https://user-images.githubusercontent.com/8207622/72237440-c146b280-35da-11ea-94a6-f4c1273d2ff1.png)
Finally I just put another scheme name into `user.lua` and all of a sudden it works.
![2020-01-13-075831_1366x768_scrot](https://user-images.githubusercontent.com/8207622/72237441-c146b280-35da-11ea-9882-a3c54a25157c.png)
Not quite sure what happened there. I'll have to try again on the desktop.

Also I haven't seen the halo effect on this machine yet. That may give me an opportunity to narrow it down on my end.
Differences between the machines are resolution 1366x768 vs. 2x1920x1080, possibly different wxGTK versions (not sure this matters since I'm using the binaries you ship), possibly different settings, e.g. on the Laptop long text wraps whereas I sometimes get vertical scroll bars on the desktop machine. Or it could be any other difference I can't think of at the moment. I'll try to compare this evening.

**Comment by pkulchenko** [comment]
> possibly different wxGTK versions (not sure this matters since I'm using the binaries you ship)

This shouldn't matter if you use packaged binaries, as these are compiled with GTK3.

> Not quite sure what happened there. I'll have to try again on the desktop.

I think I found some inconsistency in setting the tooltip colors. If you look at `cfg/scheme-picker.lua`, it shows the last command as `styles.calltip = styles.text`. If you don't do that, then the configured (non-default) scheme will have white tooltip background.

I pushed a fix that makes the tooltip colors the same as the main text colors by default. If you don't want this, then either set the tooltip colors explicitly or add `styles.calltip=nil` to the config (which will make the tooltip background white.

The fix is in upgrade-binaries-190d branch, so you can refresh yours and let me know the result.

**Comment by hollunder** [comment]
So for the 'halo-effect', it seems to only occur with `editor.usewrap = false`. Also from what I've seen it also seems to be connected to the vertical scroll bar. 

I'll give the fix a shot.

**Comment by pkulchenko** [comment]
Interesting on the halo effect. I'll check in the Scintilla maillist.

**Comment by pkulchenko** [comment]
It would also explain why it's not seen in the Output/Console windows, as those use `usewrap=true` by default.

**Comment by hollunder** [comment]
I think the tooltips are fixed but you broke the automatic dark color scheme for dark GTK theme.

**Comment by pkulchenko** [comment]
@hollunder, one more thing: when you narrow the window such that the halo is shown over the text, how does the text look like? Can you make a screenshot of that?

Asked the [question on the Scintilla maillist](https://groups.google.com/forum/#!topic/scintilla-interest/RHO_fFh2zK4).

**Comment by pkulchenko** [comment]
> I think the tooltips are fixed but you broke the automatic dark color scheme for dark GTK theme.

Hm, shouldn't make any difference; you mean it's not dark with the dark color scheme?

**Comment by pkulchenko** [comment]
@hollunder, I tested on macOS with the dark scheme and everything works correctly for me. Maybe you have something in the config that interferes with the default dark scheme?

**Comment by hollunder** [comment]
Yes, the editor is no longer dark with no editor color scheme selected and dark GTK color scheme.

Regarding how the text looks under the halo, that seems impossible to say.
1. The halo shrinks when I shrink the window.
2. When a line that is too long is shown the vertical scroll bar appears and the halo disappears.
3. The halo shrinks when there is a very long line. (screenshot in a second)

**Comment by hollunder** [comment]
On both of my machines with the latest version I get bright text with a dark GTK theme if I have no style set. Even with empty config files.

**Comment by pkulchenko** [comment]
Thank you for the halo screenshots; will update the maillist thread.

> On both of my machines with the latest version I get bright text with a dark GTK theme if I have no style set. Even with empty config files.

Can you checkout the previous commit (7fcf9eda) and let me know if this fixes the default color scheme? I don't see how the last commit could have affected that, so would like to confirm that it's indeed the last change and not one of the earlier commits.

Just in case: does your `upgrade-binaries-190d` branch show the last commit as the merge commit or was the pull applied cleanly? If there is a merge commit, you may want to delete this branch locally and then run `git fetch origin` to retrieve the clean branch.

**Comment by hollunder** [comment]
Ah, the color scheme not being set was my bad. Previously I had downloaded a zip, which gave me the correct branch. Just earlier I cloned the repo which doesn't give you the branch selected on the website. So I was using master by accident. After switching to  upgrade-binaries-190d the text style works as expected.

**Comment by pkulchenko** [comment]
Oh, good; thank you for checking and confirming! Yes, when you clone the repo, you still need to check out the branch, even when it's shown as the default on github.

**Comment by pkulchenko** [comment]
@hollunder, [from the author of the Scintilla component](https://groups.google.com/d/msg/scintilla-interest/RHO_fFh2zK4/5fXfBxbiAgAJ) (that is used in the IDE): "GTK may, depending on theme, indicate scrollability with a fade or similar effect. The searchable term for this is “undershoot”. There is also “overshoot” when trying to scroll past an end.  It could be an indication of whether the window is horizontally scrollable and to what extent. The horizontal scrollbar may still have an idea of scrollwidth even though it is not visible."

**Comment by hollunder** [comment]
Unfortunately that does not tell me anything useful. Since this appears on almost every dark theme I tried I have to assume that what we are seeing is the default behavior.
I found the mentioned terms in the GTK3 docs https://developer.gnome.org/gtk3/stable/GtkScrolledWindow.html but not in the corresponding GTK2 docs https://developer.gnome.org/gtk2/stable/GtkScrolledWindow.html.
I also glanced over the `HighContrastInverse/gtk-2.0/gtkrc` but could not find anything that appeared relevant. Most of the dark themes I tested have a GTK2 as well as a GTK3 version.

All I do know for certain is that this does not appear with zbs 1.80 or `master` with the exact same settings. I tried all the 190 branches and they all show the issue, so for now I assume that it is due to some upgraded binaries rather than zbs code.

In fact I could narrow it down to this commit: https://github.com/pkulchenko/ZeroBraneStudio/commit/7c9bd0754d8541d9310dba3f4895d5bbad04aa8d

**Comment by pkulchenko** [comment]
Right; this is the commit that included upgraded binaries. These binaries are compiled against GTK3, whereas the 1.80 and older were compiled against GTK2, so this indeed could be related to some GTK3-specific behavior. If that's the case, there is nothing that Scintilla/wxwidgets/ZBS can do to control this.

If this bothers you too much, we can probably build one version with GTK2, but I still plan to use GTK3 in the shipped binaries, as it's newer and better supported.

**Comment by pkulchenko** [comment]
That would be great; thank you for looking into this!

**Comment by hollunder** [comment]
Just a heads up.
I had the situation that the 'halo' no longer appeared on one of my machines. Unfortunately I didn't have time to investigate that thoroughly. Right now it seems to be gone on both of my machines.
I don't know what's up, but I suspect some weirdness regarding GTK2/GTK3 theming, but I can't say for certain. Just for future reference, I switched themes using lxappearance, the GTK2 and GTK3 version respectively. I can't rule out that that did something weird either.

Anyway, I certainly won't have time to look into it anytime soon and besides, the issue seems to be gone for now. Don't let this stop you from releasing 1.90 (I suspect it wouldn't, but just in case).

**Comment by pkulchenko** [comment]
> I don't know what's up, but I suspect some weirdness regarding GTK2/GTK3 theming, but I can't say for certain. Just for future reference, I switched themes using lxappearance, the GTK2 and GTK3 version respectively. I can't rule out that that did something weird either.

@hollunder, thank you very much for the update! Don't hesitate to add a comment to this ticket if you figure out what's going on with the halo or have some additional details.

**Comment by pkulchenko** [comment]
This has been implemented as part of the changes in #999 that have been merged.

**Comment by moteus** [comment]
I just update ZBS on my Ubunty 18.04 and got the same `halo` effect. It appears in my case when I set `editor.usewrap = false` in my config. Also, this effect presents on files without long lines and it dissapera if I type long line. Is there any workaround about this ?

**Comment by pkulchenko** [comment]
@moteus, I don't think so, as it doesn't seem to be controlled from the IDE or Scintilla. I included an answer from Scintilla author to my question in the earlier comment (https://github.com/pkulchenko/ZeroBraneStudio/issues/1024#issuecomment-573912989). Does it only happen on the dark background or do you see it with the default color scheme as well?

**Comment by moteus** [comment]
I see the same on all themes. But it noticeable only on dark one. With the light background it almost not seen.

**Comment by ildar** [comment]
Am i getting it right that wx should be patched? Otherwise tabs labels are unreadable?

**Comment by pkulchenko** [comment]
@ildar, no, I applied a patch to wxwidgets that is supposed to fix the contrast on tab labels with a dark color scheme.

Just to confirm, you are using v1.90, right? If you see low contrast on tab labels with the dark mode selected, can you post a screenshot?

**Comment by ildar** [comment]
I'm sure everything is great with ZBS bundle. But i use system binaries.
So i wanted to understand the way to fix it:
1. Is there a patch for Lua sources OR
2. It's done with patching WX sources OR
3. I need to upgrade to 3.1.3 OR
4. 3 and 2 together?
Thanks

**Comment by pkulchenko** [comment]
> I need to upgrade to 3.1.3

This should be sufficient. Unfortunately odd version number is  a "dev" version, so may not be available as a system binary. You can definitely take my patch and apply it yourself, but it can be a bit messy, as the changes were in several PRs/updates. See this ticket for the discussion and all related patches: https://trac.wxwidgets.org/ticket/18601

Another problem with 3.1.3 was that there were several serious defects that were fixed shortly after it was released and incorporated into ZBS binaries (the IDE reports 3.1.4 as the version if you check the About screen). So, if you have a choice, I'd recommend building from sources in pkulchenko/wxWidgets instead of wxWidgets/wxWidgets. I don't remember the nature of the defects, but can dig it up if you are interested.

**Comment by moteus** [comment]
Regarding halo effect. I turn it off using [this answer](https://ubuntu-mate.community/t/how-to-disable-this-effect/18929/2)

**Comment by pkulchenko** [comment]
@moteus, great; thank you for the update! I suspect it had something to do with GTK settings, so it's good to have a confirmation. Just in case, here is the solution:

```
overshoot.top,
overshoot.right,
overshoot.bottom,
overshoot.left {
  background: none;
}

(could also add for undershoot)

undershoot.top,
undershoot.right,
undershoot.bottom,
undershoot.left,
```

**Comment by ildar** [comment]
On Fri, Apr 17, 2020 at 12:26 PM Paul Kulchenko wrote:

> I need to upgrade to 3.1.3
>
> This should be sufficient.
>
I just found it's not.
> ZBS About page: Built with wxWidgets 3.1.3, wxLua 3.0.0.8, Scintilla
3.7.2

Testing showed:
1. Tab labels are dark
2. Text scaling factor is not considered
3. Autoswitching to the dark theme works.

**Comment by pkulchenko** [comment]
@ildar, just to confirm, these are the libraries your built, right? Does 1 and 2 work with the ones included with the IDE (3 works in both places)? I think 1 was included after 3.1.3 was released and 2 is probably result of a bug in wxwidgets that I mentioned (that also got resolved shortly after the release was made). That's why I suggested building from pkulchenko/wxwidgets master branch, as it includes those changes on top of 3.1.3 (and will report the version as 3.1.4). You can at least check the last commit there, as it will have all those fixes already included.

**Comment by ildar** [comment]
You got it all right. Thank you

**Comment by hollunder** [comment]
Thanks, I can confirm this works with the official zbstudio build.
Just to make it more obvious for others how to apply that workaround.

`~/.config/gtk-3.0/gtk.css`
```css
overshoot.top,
overshoot.right,
overshoot.bottom,
overshoot.left,
undershoot.top,
undershoot.right,
undershoot.bottom,
undershoot.left
{
	background: none;
}
```

**Comment by cata0309** [comment]
For some reason the gtk.css workaround doesn't work for me. I still get the overshot effect and I just hate it.
![image](https://user-images.githubusercontent.com/35623921/84590528-262f5e80-ae40-11ea-9fb0-ae522d0ed7f0.png)

**Comment by pkulchenko** [comment]
@cata0309, it's not overshoot; it's the currently selected tabart in the IDE. You may switch to a different one (that doesn't have the gradient) by adding the following to the config: `wxaui.wxAuiGenericTabArt = wxaui.wxAuiSimpleTabArt` or `wxaui.wxAuiGenericTabArt = wxaui.wxAuiDefaultTabArt`.

**Comment by cata0309** [comment]
Could you be more specific ? In what file do I need to edit?

**Comment by cata0309** [comment]
got it ! It was `/opt/zbstudio/zbstudio/config.lua` Now my question is how to make the tabs another color other than white ?

**Comment by pkulchenko** [comment]
> got it ! It was /opt/zbstudio/zbstudio/config.lua

That's not the place, as this file will be overwritten upon upgrade. See the Configuration section in the documentation: https://studio.zerobrane.com/doc-configuration#system-wide-configuration

> Now my question is how to make the tabs another color other than white ?

These are set based on the system colors, but there is a way to change the default. Adding something like this to the config will change the tab background to Blue and active tab to Green:

```lua
local tabart = wxaui.wxAuiSimpleTabArt()
tabart:SetColour(wx.wxColour(0, 0, 255))
tabart:SetActiveColour(wx.wxColour(0, 255, 0))
getmetatable(wxaui.wxAuiGenericTabArt).__call = function() return tabart end
```

You can also change the fonts according to the documentation here: https://docs.wxwidgets.org/trunk/classwx_aui_tab_art.html

**Comment by mimi89999** [comment]
I think you should see some improvement soon. See https://github.com/wxWidgets/wxWidgets/pull/1950 and tell me if you like it.

**Comment by cata0309** [comment]
What does `soon` mean in this context ? It could be a week/month/year or maybe more.

**Comment by pkulchenko** [comment]
@mimi89999, the change you are proposing is not relevant to what @cata0309  is discussing, as it doesn't change the background. It's only supposed to change the color of text on tabs (and it doesn't work correctly when the dark/light scheme is switched, as commented in the ticket).

If anything, what's being discussed in https://trac.wxwidgets.org/ticket/18812 may address @cata0309's concern. I'm not sure when it's going to be implemented and integrated.

**Comment by mimi89999** [comment]
> What does soon mean in this context ? It could be a week/month/year or maybe more.

I hope that the PR will be merged in the following days. https://trac.wxwidgets.org/wiki/Roadmap says _we plan to release one last 3.1.4 release in 3.1.x series in May or June 2020_

> It's only supposed to change the color of text on tabs (and it doesn't work correctly when the dark/light scheme is switched, as commented in the ticket).

No. It changes the color of the background behind tabs as can be clearly seen on the screenshots.

**Comment by pkulchenko** [comment]
> I hope that the PR will be merged in the following days. https://trac.wxwidgets.org/wiki/Roadmap says we plan to release one last 3.1.4 release in 3.1.x series in May or June 2020

But this is only for the wxwidgets component; you'll have to recompile the binaries for the IDE itself to get the update.

> No. It changes the color of the background behind tabs as can be clearly seen on the screenshots.

It does, but it's not part of this issue that we are discussing; the main issue was that the text was not visible on tabs, which has already been addressed. And this particular change (with the background) was done intentionally, for the reasons I outlined in my comments for https://trac.wxwidgets.org/ticket/18812. I don't mind changing this (as I commented in the ticket), but I don't think your proposal is the way to go, as it has some deficiencies that have already been noted in the ticket.
